### PR TITLE
♻️ [RUMF-1368] use the PointerDown event target for click actions

### PR DESCRIPTION
**Pull request by BenoitZugmeyer**
### Click target inaccuracy

When the click starts (mouse/pointer down) on an element and ends (mouse/pointer up) on another element, the `click` event target is set to the `BODY` element by the browser. This is problematic, because we report  clicks occuring on the `BODY` even if the user uses a whole other element. This impacts the RUM click action target name (`@action.target.name`), as well as the experimental heatmaps CSS selector generation.

Using the `pointerdown` event target instead improves the situation: the element used to compute the action target name and CSS selector is the element the user intended to click.

### Heatmap CSS selector improvement

Some application add a temporary class name when an element is being clicked (similarly to the [`:active`](https://developer.mozilla.org/en-US/docs/Web/CSS/:active) CSS pseudo-class) or focused. Those temporary class names blocks us from finding again the element when rendering the Heatmap again.

Generating the CSS selector before the application UI reacts to it should improve the situation a bit (`pointerdown` is the first event triggered when a click occurs).

### Frustration signal follow-up

This change is also a first step in fixing one of the most common dead click false positive cases: currently, we are only watching for page activity during or following a `click` event. But in some cases the application updates the UI earlier (ex: on `mousedown`, `focus`...). Those kind of clicks we are incorrectly considered a click as "dead".

In a future follow-up, we'll fix this by watching page activity during or following the `pointerdown` event as well, to avoid marking those clicks as "dead".

## Changes

Beside a bit of refactoring, the main change was to update the `listenActionEvents` to call a callback on `pointerdown`. This callback can return a `onClick` closure that will be called during the next `click` event. This allows to build the target-related properties earlier and still build the `Click` during the `click` event.

## Testing

- [x] Local
- [x] Staging
- [x] Unit
- [x] End to end

## Notes

This PR does not fix the issue in the recorder part:  clicks event will still be reported on the `body` element when the element target changes between mousedown and mouseup. This should not matter too much, as the click position is also collected.

To keep backward compatibility, the `event` send as context in the `beforeSend` callback is still the `click` event.

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: packages/rum-core/src/domain/rumEventsCollection/action/trackClickActions.ts [hunk 1]
```diff
@@ -95,79 +105,97 @@ export function trackClickActions(
...
+  const clickActionBase = computeClickActionBase(pointerDownEvent, configuration.actionNameAttribute)
+  if (!configuration.trackFrustrations && !clickActionBase.name) {
+    // TODO: remove this in a future major version. To keep retrocompatibility, ignore any action
+    // with a blank name
+    return
+  }
 
-    const click = newClick(lifeCycle, history, getUserActivity, clickActionBase)
+  return {
+    onClick({ event: clickEvent, getUserActivity }: OnClickContext) {
```
[the remark this answers is not in the document]
Comment (BenoitZugmeyer) [reply packages/rum-core/src/domain/rumEventsCollection/action/trackClickActions.ts:147]: > **💭 thought:** ‏IMO the current API is a bit hard to follow especially the optional onClick to indicate that a click has been aborted.
> 
> Could we keep all the processing at click level but provides both pointerdown and click events to the processing? listenActionEvents would only be responsible to keep this event and provide it later.

No, as we want to compute the CSS selector before the UI reacts to the click (second motivation in the PR description 🙂 )

> 
> Otherwise, what about splitting the processing of the pointer down from the processing of the click? We could rely on the result of the former to consider the click as aborted or not.

I'll make something.

File: packages/rum-core/src/domain/rumEventsCollection/action/listenActionEvents.ts [hunk 2]
```diff
@@ -57,6 +62,7 @@ export function listenActionEvents({ onClick }: { onClick: OnClickCallback }) {
             event: clickEvent,
             getUserActivity: () => userActivity,
           })
+          onClick = undefined
```
[the remark this answers is not in the document]
Comment (BenoitZugmeyer) [reply packages/rum-core/src/domain/rumEventsCollection/action/listenActionEvents.ts:65]: We don't want to call the `onClick` callback multiple times. This can happen for example when the application is triggering a `click` event.

I'll add a test for this.

[the remark this answers is not in the document]
Comment (bcaudan) [reply packages/rum-core/src/domain/rumEventsCollection/action/listenActionEvents.ts:65]: to clarify my understanding, it is not related to the current change and could already happen, right?

**Review by liywjl — APPROVED** [review]

File: packages/rum-core/src/domain/rumEventsCollection/action/listenActionEvents.spec.ts [hunk 3]
```diff
@@ -1,27 +1,66 @@
...
+      onPointerDown: jasmine.createSpy().and.returnValue({}),
+    }
+    ;({ stop: stopListenEvents } = listenActionEvents(actionEventsHooks))
   })
 
   afterEach(() => {
     stopListenEvents()
   })
 
+  it('listen to mousedown events', () => {
```
Comment (bcaudan) [inline packages/rum-core/src/domain/rumEventsCollection/action/listenActionEvents.spec.ts:25]:
```suggestion
  it('listen to pointerdown events', () => {
```

File: packages/rum-core/src/domain/rumEventsCollection/action/trackClickActions.ts [hunk 4]
```diff
@@ -79,11 +78,25 @@ export function trackClickActions(
   lifeCycle.subscribe(LifeCycleEventType.BEFORE_UNLOAD, stopClickChain)
   lifeCycle.subscribe(LifeCycleEventType.VIEW_ENDED, stopClickChain)
 
-  const { stop: stopActionEventsListener } = listenActionEvents({ onClick: processClick })
+  const { stop: stopActionEventsListener } = listenActionEvents<ClickActionBase>({
+    onPointerDown: (pointerDownEvent) => onPointerDown(configuration, history, pointerDownEvent),
+    onClick: (clickActionBase, clickEvent, getUserActivity) =>
+      onClick(
```
Comment (bcaudan) [inline packages/rum-core/src/domain/rumEventsCollection/action/trackClickActions.ts:84]: **💬 suggestion:** ‏to differentiate the hook from the performed action
```suggestion
  const { stop: stopActionEventsListener } = listenActionEvents<ClickActionBase>({
    onPointerDown: (pointerDownEvent) => processPointerDown(configuration, history, pointerDownEvent),
    onClick: (clickActionBase, clickEvent, getUserActivity) =>
      processClick(
```

**Review by bcaudan — APPROVED** [review]
Nice!
